### ISSUE TITLE
Warn on da.empty usage ?

### ISSUE THREAD

**Issue by Carreau** [dask/dask#6511]
See https://github.com/dask/dask/pull/6491#issuecomment-673601353 and later comments for background.

But roughly as dask array are immutable; usage of `empty()` beyond drop-in replacement of numpy are likely wrong and could/should be replaced by full/ones/zeroes which will also be more efficient once #6491 is merged.

**Comment by jsignell** [comment]
Dask arrays aren't totally immutable though. You can do things like:

```python
import dask.array as da

d = da.empty((3,5))
d[d == 0] = 3
d.compute()
```

**Comment by Carreau** [comment]
Sorry I misexpressed myself.

I don't think the memory that backs d here: `d = da.empty((3,5))` 
and _after_ `d[d == 0] = 3` is the same, is it ? 

I understand `d[d == 0] = 3` as adding a step to the graph; so the definition of "After" is weird, but still; I don't think the memory location of the original empty can be modified. 

So If my understanding is correct I don't believe there implementations that can rely on creating a `empty()` array for memory efficiency, keep reference to it and then mutate the internal memory state, they will end up allocating anyway. And in the case like you show above, comparing values of an `empty()` array to a set value is recipe for bugs, so they should use `full(Nan)`, `ones()`, or `zeros()`

What make we think that the memory is untouched is the following master, ones/zeros are backed by broadcast trick and this underlying data store should be immutable (x.flags.writable == 0 in numpy land), though you can still do:


```
import dask.array as da
d = da.zeros((3,5))
d[d == 0] = 3
```

And this works well. If this was a real assignment, it would raise an error. 

Is that a better explanation of what I am trying to express ?

**Comment by jsignell** [comment]
Thanks for explaining that. I think what you are saying makes sense. And agree that I don't know of a time when empty is a better choice than zeros.

**Comment by jcrist** [comment]
I'm not sure we should warn on `empty`, as it's nice to have it for full compatibility with the numpy api. And since they're all backed by stride tricks now, I'm not sure if there's really a cost to using `da.empty` this way? I'd opt to close this and leave things as is.

**Comment by jakirkham** [comment]
Yeah that seems ok to me as well. Increasingly the community is working on making sure different arrays can be used easily in downstream code. So having warnings pop-up for some APIs might confuse people.

That said, I understand what Matthias is getting at. Most of the reason people use `np.empty` (or similar) is to do an `malloc` instead of a `calloc` for example (avoiding the cost of initializing that array). This is a weird thing to do in a lazy array library and could cause issues if these uninitialized values are used in downstream computations (often if not always a mistake?). Idk if warning is the answer (as mentioned above), but maybe documenting best practices with Arrays and/or data initialization in Dask would be helpful?
